File: leetcode/problems/problem-specific/57.insert-interval/57.insert-interval.go
package main

func insert(intervals [][]int, newInterval []int) [][]int {
	if len(intervals) <= 0 {
		return [][]int{{newInterval[0], newInterval[1]}}
	}

	insertPosition := findInsertPosition(intervals, newInterval)
	// * Switch elements to the right
	intervals = append(intervals, []int{})
	copy(intervals[insertPosition+1:], intervals[insertPosition:])
	intervals[insertPosition] = newInterval

	return merge(intervals)
}

func findInsertPosition(intervals [][]int, newInterval []int) int {
	i, n := 0, len(intervals)
	// To handle TC --> intervals = [[1,4]], newInterval = [0,5]
	if n <= 1 {
		if newInterval[0] >= intervals[0][0] {
			return 1
		}

		return 0
	}

	for i = 0; i < n-1; i++ {
		// To handle TC --> intervals = [[1,5],[6,8]], newInterval = [0,9]
		if i == 0 && newInterval[0] <= intervals[i][0] {
			return i
		} else if newInterval[0] >= intervals[i][0] && newInterval[0] < intervals[i+1][0] {
			break
		}
	}

	return i + 1
}

func merge(intervals [][]int) [][]int {
	if len(intervals) <= 1 {
		return intervals
	}

	res, n, i := [][]int{}, len(intervals), 0
	for i = 0; i < n-1; i++ {
		start, end := intervals[i][0], intervals[i][1]
		for i < n-1 && end >= intervals[i+1][0] {
			end = findMax(end, intervals[i+1][1])
			i++
		}

		res = append(res, []int{start, end})
	}

	if i == n-1 {
		res = append(res, intervals[i])
	}

	return res
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/problem-specific/57.insert-interval/57.insert-interval.py
#
# @lc app=leetcode id=57 lang=python3
#
# [57] Insert Interval
#

# @lc code=start

# * Iterative Solution | O(n) Time | O(n) Space
# * n -> The length of intervals array


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        inserted_intervals = []
        for idx, interval in enumerate(intervals):
            # * No Overlap Case.

            # * Insert before the current interval.
            # * Add the newInterval to the inserted intervals and return the inserted
            # * intervals appended with the remaining intervals if the end of the
            # * newInterval is less than the start of the current interval.
            if newInterval[1] < interval[0]:
                inserted_intervals.append(newInterval)
                return inserted_intervals + intervals[idx:]
            # * Insert after the current interval.
            # * Add the current interval to the inserted intervals if the start of
            # * the newInterval is greater than the end of the current interval.
            elif newInterval[0] > interval[1]:
                inserted_intervals.append(interval)

            # * Overlap Case.
            else:
                newInterval = [
                    min(newInterval[0], interval[0]),
                    max(newInterval[1], interval[1]),
                ]

        inserted_intervals.append(newInterval)
        return inserted_intervals


# @lc code=end
